tools/make_patch.c

```diff
@@ -0,0 +1,891 @@
+	}
+
+	return NULL;
+}
+
+int parse_address_bank(char *buffer_input) {
+	int bank;
+	char *buffer = malloc(strlen(buffer_input));
+	char *endptr_bank;
+	char *buffer_address;
+
+	strcpy(buffer, buffer_input);
+	buffer_address = strchr(buffer, ':');
+	if (buffer_address) *buffer_address++ = '\0';
+
+	// Verify it consists of expected elements
+	if (strlen(buffer) > 2 || !buffer_address || strlen(buffer_address) > 4) {
+		if (buffer_address) *--buffer_address = ':';
+		fprintf(stderr, "Error: Unexpected %s\n", buffer);
+		return -1;
+	}
+
+	// Parse the bank part
+	bank = strtol(buffer, &endptr_bank, 16);
+	if (endptr_bank != buffer + strlen(buffer)){
+		fprintf(stderr, "Error: Cannot parse bank %s:%s\n", buffer, buffer_address);
+		return -1;
+	}
+
+	return bank;
+}
+
+int parse_address_address(char *buffer_input) {
+	int address;
+	char *buffer = malloc(strlen(buffer_input));
+	char *endptr_address;
+	char *buffer_address;
+
+	strcpy(buffer, buffer_input);
+	buffer_address = strchr(buffer, ':');
+	if (buffer_address) *buffer_address++ = '\0';
+
+	// Verify it consists of expected elements
+	if (strlen(buffer) > 2 || !buffer_address || strlen(buffer_address) > 4){
+		if (buffer_address) *--buffer_address = ':';
+		fprintf(stderr, "Error: Unexpected %s\n", buffer);
+		return -1;
+	}
+
+	// Parse the address part
+	address = strtol(buffer_address, &endptr_address, 16);
+	if (endptr_address != buffer_address + strlen(buffer_address)){
+		fprintf(stderr, "Error: Cannot parse bank %s:%s\n", buffer, buffer_address);
+		return -1;
+	}
+
+	return address;
+}
+
+int parse_abs_offset(int bank, int address, char type) {
+	switch (type) {
+	case('w'):
+		return bank ? bank * 0xd000 + (address - 0xd000) : address;
+	case('s'):
+		return bank ? bank * 0xc000 + (address - 0xc000) : address;
+	case('h'):
+		return address;
+	default:
+		return bank ? bank * 0x4000 + (address - 0x4000) : address;
+	}
+
+	return -1;
+}
+
+int parse_offset(int offset, char type) {
+	switch(type) {
+	case('w'):
+		return offset > 0xd000 ? 0xd000 + offset % 0xd000 : offset;
+	case('s'):
+		return offset > 0xc000 ? 0xc000 + offset % 0xc000 : offset;
+	case('h'):
+		return offset;
+	default:
+		return offset > 0x4000 ? 0x4000 + offset % 0x4000 : offset;
+	}
+	
+	return -1;
+}
+
+struct symbol *parse_symfile(FILE *file) {
+	struct symbol *symbols = NULL;
+	char *buffer;
+	size_t buffer_index = 0;
+	int c;
+	int bank = 0;
+	int address = 0;
+	int offset = -1;
+
+	buffer = create_buffer();
+	if (!buffer) goto error;
+
+	// Parse the entire symfile, to gather the symbols
+	while ((c = getc(file)) != EOF) {
+		buffer = expand_buffer(buffer, buffer_index);
+		if (!buffer) goto error;
+
+		switch (c) {
+		case ';':
+		case '\n':
+			// If we encounter a newline or comment,
+			// we've reached the end of the label name.
+			buffer[buffer_index] = '\0';
+
+			// This is the end of the parsable line
+			buffer_index = SIZE_MAX;
+			offset = parse_abs_offset(bank, address, buffer[0]);
```

Comment:
This is relying on pokecrystal to consistently name its labels with the `w`, `h`, and `s` prefixes (and `v`, although `parse_offset` and `parse_abs_offset` don't handle that). Instead it should use the value of the address to decide whether it's in SRAM, HRAM, etc.